### PR TITLE
feat(build-a-box): Add example value box builder app

### PR DESCRIPTION
**Pull request by gadenbuie**
image

For #778

To run the app call the following

```r
shiny::runApp(system.file("examples/build-a-box", package = "bslib"))
```

There's only one external "option" that is controlled by the `ENABLE_THEMER` envvar. When exactly `"true"`, the themer is included in the app -- it's also minimized on init with a bit of JavaScript (maybe we should add first-class support for controlling how or where the themer UI is inserted).

```r
Sys.setenv("ENABLE_THEMER" = "true")
shiny::runApp(system.file("examples/build-a-box", package = "bslib"))
```

You can also turn on the themer for a running app by adding `?themer` or `?themer=true` to the URL and reloading the page.

### CODE REVIEW

File: inst/examples/build-a-box/app.R [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+library(plotly)
+
+# Settings ---------------------------------------
+ENABLE_THEMER <- identical(Sys.getenv("ENABLE_THEMER"), "true")
+
+# Functions ---------------------------------------
+layout_value_box_options <- function(ui_opts) {
+  layout_columns(
+    gap = "1rem",
```
Comment (cpsievert) [inline inst/examples/build-a-box/app.R:22]: Not sure if this is necessary

File: inst/examples/build-a-box/www/build-a-box.css [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/* FIXME: Remove after rstudio/htmltools#401 */
+.html-fill-container > .html-fill-item {
+  overflow: visible;
+}
```
Comment (cpsievert) [inline inst/examples/build-a-box/www/build-a-box.css:4]: Has been merged to main, so could remove (but make sure to add version requirement & `Remotes`)

File: inst/examples/build-a-box/app.R [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    div(h4("Content", class = "border-bottom"), ui_opts$title_value),
+    div(h4("Theme", class = "border-bottom"), ui_opts$theme_opts),
+    div(h4("Showcase", class = "border-bottom"), ui_opts$showcase)
+  )
+}
+
+# UI ---------------------------------------
+ui <- page_fixed(
+  title = "Build a Box | bslib",
+  theme = bs_theme(preset = "shiny"),
```
Comment (cpsievert) [inline inst/examples/build-a-box/app.R:32]:
```suggestion
  theme = bs_theme(preset = "shiny") %>% bs_add_rules(sass::sass_file("build-a-box.scss")),
```

File: inst/examples/build-a-box/R/code_modal.R [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        class = "text-muted d-none",
+        HTML("Press <kbd>Ctrl</kbd>/<kbd>Cmd</kbd> + <kbd>C</kbd> to copy the value box example code.")
+      ),
+      tags$button(
+        id = "copy-code-to-clipboard",
+        class = "btn btn-outline-primary",
+        onclick = "copyValueBoxCode()",
+        "Copy to clipboard"
+      ),
+      tags$script(HTML("checkCopyPermissions()")),
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/code_modal.R:29]: Add a comment here about how this assumes `build-a-box.js` has already been sources

Comment (cpsievert) [reply inst/examples/build-a-box/R/code_modal.R:29]: That or pull the JS into a separate file that comes with this component

File: inst/examples/build-a-box/R/mod-global-controls.R [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        "Showcase Layout",
+        c("", "left center", "top right", "bottom"),
+        "left center"
+      )
+    )
+  )
+}
+
+server_global_controls <- function(input, output, sessions, one, two, three) {
+  observeEvent(input$random_theme, {
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-global-controls.R:60]: Currently, if you choose a new theme style, the value boxes don't update, and it's not necessarily obvious you need to click random theme to apply the new style

```suggestion
  observeEvent({input$random_theme, input$theme_style}, {
```

File: inst/examples/build-a-box/R/mod-global-controls.R [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          ns = ns,
+          shuffleButton(ns("random_icon"), "Icons")
+        )
+      ),
+      selectInput(
+        ns("showcase_layout"),
+        "Showcase Layout",
+        c("", "left center", "top right", "bottom"),
+        "left center"
+      )
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-global-controls.R:54]: In each of the value box tabs, `radioButtons()` are used to choose these options. I don't have a strong preference either way, but probably best to be consistent?

File: inst/examples/build-a-box/R/mod-value-box-ui.R [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+      textInput(ns("title"), "Title", init$title),
+      textInput(ns("value"), "Value", init$value),
+      tagAppendAttributes(
+        textAreaInput(ns("extra"), "Extra text (markdown allowed)", ""),
+        class = "input-text-code"
+      ),
+      actionButton(ns("random_stat"), "Random stat")
+    ),
+    "theme_opts" = list(
+      input_switch(ns("use_theme"), "Use theme", TRUE),
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-value-box-ui.R:50]: I think the UX would be a tad more intuitive if it was more like this:

```r
input_switch(ns("use_custom_colors"), "Use custom colors", FALSE),
```

File: inst/examples/build-a-box/R/mod-global-controls.R (hunk 5 above)
Comment (gadenbuie) [reply inst/examples/build-a-box/R/mod-global-controls.R:60]: Thanks I noticed this too. It needs to be `c(input$random_theme, input$theme_style)` though and also requires `ignoreInit = TRUE`

File: inst/examples/build-a-box/R/mod-value-box-ui.R [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+          min = 100,
+          max = 500,
+          value = 150,
+          post = "px",
+          step = 10
+        )
+      )
+    ),
+    showcase = list(
+      input_switch(ns("showcase"), "Showcase", value = TRUE),
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-value-box-ui.R:77]:
```suggestion
      input_switch(ns("showcase"), "Include showcase", value = TRUE),
```

File: inst/examples/build-a-box/R/mod-value-box-ui.R [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+        ui_selextra(ns("theme"), "Theme")
+      ),
+      conditional_panel(
+        "!input.use_theme",
+        colourpicker::colourInput(ns("foreground"), "Foreground", value = "#000000"),
+        colourpicker::colourInput(ns("background"), "Background", value = "#FFFFFF")
+      ),
+      input_switch(ns("full_screen"), "Allow full screen", value = FALSE),
+      input_switch(ns("fill"), "Fill vertical space", value = TRUE),
+      input_switch(ns("fixed_height"), "Fixed height", value = FALSE),
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-value-box-ui.R:62]: It kind of feels like these could/should be global controls

File: inst/examples/build-a-box/R/mod-global-controls.R [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        )
+      ),
+      div(
+        class = "btn-group",
+        role = "group",
+        "aria-label" = "Theme actions",
+        shuffleButton(ns("random_theme"), "Theme"),
+        shuffleButton(ns("random_stat"), "Stats"),
+        actionButton(ns("reset_theme"), "Reset")
+      )
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-global-controls.R:32]: I think you could get away with not having this button group if you wanted. I'm not sure the Stats/Reset is needed and you could put a refresh icon in the theme dropdown (like is done on the other tabs)?

File: inst/examples/build-a-box/R/mod-value-box-ui.R (hunk 9 above)
Comment (gadenbuie) [reply inst/examples/build-a-box/R/mod-value-box-ui.R:62]: Full screen follows the plot/icon choice and the other two could definitely be global options... but I think they'd clutter the initial UI without adding much value there.

File: inst/examples/build-a-box/R/mod-global-controls.R [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        )
+      ),
+      radioButtons(
+        ns("showcase_layout"),
+        "Showcase Layout",
+        choices = c("Left center", "Top right", "Bottom"),
+        inline = TRUE
+      )
+    )
+  )
```
Comment (cpsievert) [inline inst/examples/build-a-box/R/mod-global-controls.R:56]: I think this (meaning the global controls tab) UI, in general, could do a little more to guide the user experience. Here is a couple ideas:

1. Provide a button towards the bottom that says something like "Customize 1st value box". Without some visual cue like this I think the use might be left wondering what to do after exploring global controls.
2. Sort of in line with (1), perhaps include a tooltip on the theme dropdown that explains how this is a "high-level" theme control, and "lower-level" controls can be accessed on a particular value box pages.

File: inst/examples/build-a-box/R/mod-global-controls.R (hunk 10 above)
Comment (gadenbuie) [reply inst/examples/build-a-box/R/mod-global-controls.R:32]: I kind of like it. I've found it helpful to shuffle through the stats until I get a set that feel like what I'm looking for. And resetting is annoying to do. But I do think it makes more sense to have "Default" be a theme style; which helps the initial state of the input be more consistent with the value boxes (they start with default theme and the input says "default").

File: inst/examples/build-a-box/R/mod-global-controls.R (hunk 11 above)
Comment (gadenbuie) [reply inst/examples/build-a-box/R/mod-global-controls.R:51]: Great points. I added a "?" box next to "Value box settings" with a popover explaining how to use the settings panel.

File: inst/examples/build-a-box/app.R [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+      )
+    ),
+
+    # Settings Panel ----
+    navset_card_pill(
+      id = "settings",
+      title = span(
+        "Value box settings",
+        popover(
+          bsicons::bs_icon("question-square-fill"),
```
Comment (cpsievert) [inline inst/examples/build-a-box/app.R:99]: I don't feel super strongly about this, but I thikn it'll work better in dark mode

```suggestion
          bsicons::bs_icon("question-square"),
```

**Comment by cpsievert** [comment]
Probably worth adding `rstudio/shiny` to the `Config/Needs/deploy` field in `DESCRIPTION` (so the dropdowns work better in dark mode)

**Review by cpsievert — APPROVED** [review]
So awesome, feel free to merge after the minor suggestions!

**Comment by cpsievert** [comment]
One other nit. I think `input_dark_mode()` could come out the card and go in the page header (and maybe the same for the code export?)

**Comment by gadenbuie** [comment]
> One other optional nit. I think `input_dark_mode()` could come out the card and go in the page header (and maybe the same for the code export?)

I like this, thanks for the suggestion!

image

File: inst/examples/build-a-box/app.R (hunk 12 above)
Comment (gadenbuie) [reply inst/examples/build-a-box/app.R:109]: I'm going to stick with the filled version, thanks!

**Pull request merged**